// @ghostery/url-parser/dist/commonjs/search-params-wrapper.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const url_search_params_js_1 = __importDefault(require("./url-search-params.js"));
/**
 * Wraps URLSearchParams and pushes changes in search string to a parent
 * URL instance.
 */
class URLSearchParamsWrapper extends url_search_params_js_1.default {
    constructor(url, init) {
        super();
        this.url = url;
        this.params = init.params;
    }
    append(name, value) {
        super.append(name, value);
        this.url.search = this.toString();
    }
    delete(name) {
        super.delete(name);
        this.url.search = this.toString();
    }
    set(name, value) {
        super.set(name, value);
        this.url.search = this.toString();
    }
    sort() {
        super.sort();
        this.url.search = this.toString();
    }
}
exports.default = URLSearchParamsWrapper;
